Once KafkaProducer reports "Failed to allocate memory within the configured" exceptions, it never recovers from the failures

We have a situation there is a message burst in our business codes using Kafka 1.3.2. The message rate might reach 100K/s, each message size is less than 16Kb. The producer is shared between threads and configured as default parameters except 'compression_type='gzip', linger_ms=5, retries=3'.
We observe that during message burst, Kafka producer reports:
“KafkaTimeoutError: KafkaTimeoutError: Failed to allocate memory within the configured max blocking time”
And it never comes back.
Traceback listed below:
Traceback (most recent call last):
  File "/root/search/db_publisher.py", line 47, in publish
    self._publish(message)
  File "/root/search/db_publisher.py", line 41, in _publish
    self.producer.send(self.topic, msg)
  File "/usr/lib/python2.7/site-packages/kafka/producer/kafka.py", line 511, in send
    self.config['max_block_ms'])
  File "/usr/lib/python2.7/site-packages/kafka/producer/record_accumulator.py", line 243, in append
    buf = self._free.allocate(size, max_time_to_block_ms)
  File "/usr/lib/python2.7/site-packages/kafka/producer/buffer.py", line 201, in allocate
    "Failed to allocate memory within the configured"
KafkaTimeoutError: KafkaTimeoutError: Failed to allocate memory within the configured max blocking time

Then I check more and find another exception produced:
Traceback (most recent call last):  
File "/root/search/db_publisher.py", line 47, in publish    
self._publish(message)  
File "/root/search/db_publisher.py", line 41, in _publish    
self.producer.send(self.topic, msg)  
File "/usr/lib/python2.7/site-packages/kafka/producer/kafka.py", line 511, in send    self.config['max_block_ms']) 
File "/usr/lib/python2.7/site-packages/kafka/producer/record_accumulator.py", line 243, in append    
buf = self._free.allocate(size, max_time_to_block_ms) 
File "/usr/lib/python2.7/site-packages/kafka/producer/buffer.py", line 207, in allocate
assert removed is more_memory, 'Wrong condition'
AssertionError: Wrong condition

I check the source codes of above exceptions in code buffer.py (class SimpleBufferPool).
There is something wrong in the situation when producer fails to allocate memory from buffer pool within time limit:
                    if self._free:                                                                                                                                      
                        buf = self._free.popleft()                                                                                                                      
                    else:                                                                                                                                               
                        raise Errors.KafkaTimeoutError(                                                                                                                 
                            "Failed to allocate memory within the configured"                                                                                           
                            " max blocking time") 
Should we popleft the semaphore appended to the waiters before raising Errors.KafkaTimeoutError, so next allocation action no need to wait for "max_time_to_block_ms" milliseconds and thus its Condition assertion can pass? Due to each de-allocation notifies the wrong Condtion objects, each allocation action has to waste "max_time_to_block_ms" to wait the semaphore signal. As consequence, our program is slowing down rapidly and not acceptable for business.
The issue has been re-produced 3 times and I will provide debug information about memory allocation and de-allocation logging.
Anyway, thanks your guys  for this wonderful library.

--- a/kafka/producer/buffer.py
+++ b/kafka/producer/buffer.py
@@ -197,6 +197,7 @@ class SimpleBufferPool(object):
                     if self._free:
                         buf = self._free.popleft()
                     else:
+                        self._waiters.remove(more_memory)
                         raise Errors.KafkaTimeoutError(
                             "Failed to allocate memory within the configured"
                             " max blocking time")

